apis/v1alpha2/validation/gateway.go

```diff
@@ -89,3 +98,27 @@ func validateListenerHostname(listeners []gatewayv1a2.Listener, path *field.Path
 	}
 	return errs
 }
+
+// validateAddresses validates each listener address
+// if there are addresses set. Otherwise, returns no error.
+func validateAddresses(addresses []gatewayv1a2.GatewayAddress, path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	re := regexp.MustCompile(`^([a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9]\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])\/[a-zA-Z0-9]+$`)
+
+	for i, a := range addresses {
+		if a.Type == nil {
+			continue
+		}
+		_, ok := addressTypesValid[*a.Type]
+		if !ok {
+			// Found something that's not one of the upstream AddressTypes
+			// Next, check for a domain-prefixed string
+			match := re.Match([]byte(*a.Type))
+			if !match {
+				errs = append(errs, field.Invalid(path.Index(i).Child("type"), a.Type, "should either be a defined constant or a domain-prefixed string (example.com/Type)"))
+			}
+		}
+
+	}
+	return errs
+}
```

Comment:
I would like to suggest a micro-optimization here:

As written this will cause _every_ call at runtime to `validateAddresses` to compile this regex. Being static however this regexp compiles to the same thing each time. This will cause work during runtime that doesn't need to be done, instead we can just make this var a glob above the function (and document it) and use it repeatedly.

For some illustration consider the following two programs:

```go
package main

import "regexp"

func main() {
	for i := 0; i < 100000; i++ {
		match()
	}
}

func match() {
	re := regexp.MustCompile(`^([a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9]\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])\/[a-zA-Z0-9]+$`)
	_ = re.MatchString("asdf")
}
```

```go
package main

import "regexp"

func main() {
	for i := 0; i < 100000; i++ {
		match()
	}
}

var re = regexp.MustCompile(`^([a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9]\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])\/[a-zA-Z0-9]+$`)

func match() {
	_ = re.MatchString("asdf")
}
```

When compiled to assembly (e.g. `go tool compile -S main.go`) you'll see that the first program will result in the `match` call being placed in the `.main` and will be be repeatedly called in a loop:

```asm
"".main STEXT size=71 args=0x0 locals=0x10 funcid=0x0 align=0x0
        ; e.t.c. (other instructions)
	0x0014 00020 (main.go:5)	XORL	AX, AX
	0x0016 00022 (main.go:6)	JMP	44
	0x0018 00024 (main.go:6)	MOVQ	AX, "".i(SP)
	0x001c 00028 (main.go:7)	PCDATA	$1, $0
	0x001c 00028 (main.go:7)	NOP
	0x0020 00032 (main.go:7)	CALL	"".match(SB)
	0x0025 00037 (main.go:6)	MOVQ	"".i(SP), AX
	0x0029 00041 (main.go:6)	INCQ	AX
	0x002c 00044 (main.go:6)	CMPQ	AX, $100000
	0x0032 00050 (main.go:6)	JLT	24
```

conversely, the second program will call it only once in `.init`:

```asm
"".init STEXT size=86 args=0x0 locals=0x18 funcid=0x0 align=0x0
        ; e.t.c. (other instructions)
	0x0020 00032 (main.go:11)	CALL	regexp.MustCompile(SB)
```

This is because `go` has no such optimization to comprehend that the input to `MustCompile` is actually static and then try to provide any optimization for that.

This may seem fairly negligible (which is why I call it a micro-optimization) but the timing results for these two variations have a large gap:

```console
$ time ./unoptimized

real	0m0.932s
user	0m1.006s
sys	0m0.038s

$ time ./optimized

real	0m0.073s
user	0m0.074s
sys	0m0.000s

```

Given our expectation that our validation code is going to run repeatedly all over the world for years to come, seems like a good a place as any to shave off some time (that would otherwise accumulate) for the very low price of moving the variable:

```suggestion
// domainPrefixedStringRegex is a regex used in validation to determine whether
// a provided string is a domain-prefixed string. Domain-prefixed strings are used
// to indicate custom (implementation-specific) address types.
var domainPrefixedStringRegex = regexp.MustCompile(`^([a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9]\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])\/[a-zA-Z0-9]+$`)

// validateAddresses validates each listener address
// if there are addresses set. Otherwise, returns no error.
func validateAddresses(addresses []gatewayv1a2.GatewayAddress, path *field.Path) field.ErrorList {
	var errs field.ErrorList
	

	for i, a := range addresses {
		if a.Type == nil {
			continue
		}
		_, ok := addressTypesValid[*a.Type]
		if !ok {
			// Found something that's not one of the upstream AddressTypes
			// Next, check for a domain-prefixed string
			match := domainPrefixedStringRegex.Match([]byte(*a.Type))
			if !match {
				errs = append(errs, field.Invalid(path.Index(i).Child("type"), a.Type, "should either be a defined constant or a domain-prefixed string (example.com/Type)"))
			}
		}

	}
	return errs
}
```

Reply:
Thanks, that's a great explanation! To be honest, kind of annoyed at myself I didn't see it. 😄 I'll get that change in now.

Reply:
Done. I'll leave the comment unresolved though so others can easily see the explanation.